openhtf/bin/units_from_xls.py

```diff
@@ -0,0 +1,250 @@
+    @measures(units.Unit('metre per second'))
+
+OpenHTF uses UNECE unit codes internally because they are relatively complete
+and modern, and because they are recognized internationally. For full details
+regarding where we get the codes from and which units are avaiable, see the
+docstring at the top of openhtf/util/units/bin/units_from_xls.py.
+"""
+
+
+import collections
+
+
+UnitInfo = collections.namedtuple('UnitInfo', 'name code suffix')
+
+ALL_UNITS = []
+
+
+# NO_DIMENSION means that there are units set, but they cannot be expressed
+# by a known dimension (such as a ratio)
+NO_DIMENSION = UnitInfo('No dimension', 'NDL', None)
+ALL_UNITS.append(NO_DIMENSION)
+NONE = UnitInfo('None', None, None)
+ALL_UNITS.append(NONE)
+'''
+
+POST = '''
+# Convenience aliases.
+MINUTE = MINUTE_UNIT_OF_TIME
+SECOND = SECOND_UNIT_OF_TIME
+
+
+class UnitLookup(object):
+  """Facilitates user-friendly access to units."""
+  def __init__(self, lookup):
+    self._lookup = lookup
+
+  def __call__(self, name_or_suffix):
+    """Provides instantiation-like access for units module."""
+    return self._lookup[name_or_suffix]
+
+
+UNITS_BY_NAME = {u.name: u for u in ALL_UNITS}
+UNITS_BY_SUFFIX = {u.suffix: u for u in ALL_UNITS}
+
+del ALL_UNITS
+
+UNITS_BY_ALL = {}
+UNITS_BY_ALL.update(UNITS_BY_NAME)
+UNITS_BY_ALL.update(UNITS_BY_SUFFIX)
+
+Unit = UnitLookup(UNITS_BY_ALL)
+'''
+
+SHEET_NAME = 'Annex II & Annex III'
+UNIT_KEY_REPLACEMENTS = {' ': '_',
+                         ',' : '_',
+                         '.': '_',
+                         '-': '_',
+                         '/': '_PER_',
+                         '%': 'PERCENT',
+                         '[': '',
+                         ']': '',
+                         '(': '',
+                         ')': '',
+                         "'": '',
+                         '8':  'EIGHT',
+                         '15': 'FIFTEEN',
+                         '30': 'THIRTY',
+                         '\\': '_',
+                         unichr(160): '_',
+                         unichr(176): 'DEG_',
+                         unichr(186): 'DEG_',
+                         unichr(8211): '_',
+                        }
+
+
+def main(argv):
+  """Main entry point for UNECE code .xls parsing."""
+  parser = argparse.ArgumentParser(
+      description='Reads in a .xls file and generates a units module for '
+                  'OpenHTF.',
+      prog='python units_from_xls.py')
+  parser.add_argument('xlsfile', type=str,
+                      help='the .xls file to parse')
+  parser.add_argument(
+      '--outfile',
+      type=str,
+      default=os.path.join(os.path.dirname(__file__), os.path.pardir, 'util',
+                           'units.py'),
+      help='where to put the generated .py file.')
+  args = parser.parse_args()
+
+  if not os.path.exists(args.xlsfile):
+    print 'Unable to locate the file "%s".' % args.xlsfile
+    parser.print_help()
+    print 'HACKABEES', args.outfile
```

Comment:
lol oops. That's not the first nor the last time a `print 'HACKABEES'` makes it into one of my PR's accidentally.